tests/src/python/test_provider_postgres.py

```diff
@@ -1141,6 +1143,27 @@ def testCurveToMultipolygon(self):
         self.assertEqual(g.childCount(), 1)
         self.assertTrue(g.childGeometry(0).vertexCount() > 3)
 
+    def testConcurrency(self):
+        """
+        The connection pool has a maximum of 4 connections defined (+2 spare connections)
+        Make sure that if we exhaust those 4 connections and force another connection
+        it is actually using the spare connections and does not freeze.
+        This situation normally happens when (at least) 4 rendering threads are active
+        in parallel and one requires an expression to be evaluated.
+        """
+        vl = QgsVectorLayer('{conn} srid=4326 table="qgis_test".{table} (geom) sql='.format(conn=self.dbconn, table='someData'), "testgeom", "postgres")
+        self.assertTrue(vl.isValid())
+        it = vl.getFeatures()
```

Comment:
It's kinda picky - but I think you should expose a getter for CONN_POOL_MAX_CONCURRENT_CONNS and then utilise it here. Otherwise this test will become ineffective if someone bumps CONN_POOL_MAX_CONCURRENT_CONNS in future.

Reply:
Right, on the other hand, it's quite a big change "just" to get this test stable for potential future updates.

Reply:
Is it? Why not just move it to Qgis?

Reply:
Maybe it's even a parameter one should be able to tune? I can imagine it's slowing down rendering on >4 core machines?

Reply:
Moved to QgsApplication